src/libraries/Microsoft.Extensions.DependencyInjection.Abstractions/src/ActivatorUtilities.cs

```diff
@@ -32,38 +32,46 @@ public static class ActivatorUtilities
             params object[] parameters)
         {
             int bestLength = -1;
-            bool seenPreferred = false;
-
             ConstructorMatcher bestMatcher = default;
 
             if (!instanceType.IsAbstract)
             {
-                foreach (ConstructorInfo? constructor in instanceType.GetConstructors())
+                ConstructorInfo[] constructors = instanceType.GetConstructors();
+                ConstructorInfo? pereferredConstructor = null;
+                foreach (ConstructorInfo? constructor in constructors)
                 {
-                    var matcher = new ConstructorMatcher(constructor);
-                    bool isPreferred = constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute), false);
-                    int length = matcher.Match(parameters);
-
-                    if (isPreferred)
+                    if (constructor.IsDefined(typeof(ActivatorUtilitiesConstructorAttribute)))
                     {
-                        if (seenPreferred)
+                        if (pereferredConstructor is not null)
                         {
                             ThrowMultipleCtorsMarkedWithAttributeException();
                         }
-
-                        if (length == -1)
-                        {
-                            ThrowMarkedCtorDoesNotTakeAllProvidedArguments();
-                        }
+                        pereferredConstructor = constructor;
                     }
-
-                    if (isPreferred || bestLength < length)
+                }
+                if (pereferredConstructor is not null)
+                {
+                    bestMatcher = new ConstructorMatcher(pereferredConstructor);
+                    bestLength = bestMatcher.Match(parameters);
+                    if (bestLength == -1)
+                    {
+                        ThrowMarkedCtorDoesNotTakeAllProvidedArguments();
+                    }
+                }
+                else
+                {
+                    foreach (ConstructorInfo? constructor in constructors)
                     {
-                        bestLength = length;
-                        bestMatcher = matcher;
+                        var matcher = new ConstructorMatcher(constructor);
```

Comment:
```suggestion
                        if (constructor is null)
                        {
                            continue;
                        }

                        var matcher = new ConstructorMatcher(constructor);
```

Reply:
@am11 What do you think if we just replace `ConstructorInfo?` with the `ConstructorInfo`. According to documentation the `GetConstructor()` call should not return collection with nullable elements. https://docs.microsoft.com/en-us/dotnet/api/system.type.getconstructors?view=net-6.0#system-type-getconstructors

Reply:
Yes, makes sense. Based on the usage elsewhere in the repo, null is not checked. GetConstructor (singular), however, returns nullable object,